v2/sourcedb-to-spanner/src/test/java/com/google/cloud/teleport/v2/source/reader/io/transform/AccumulatingTableReaderTest.java

```diff
@@ -0,0 +1,148 @@
+import java.io.Serializable;
+import java.util.concurrent.atomic.AtomicLong;
+import org.apache.beam.sdk.testing.PAssert;
+import org.apache.beam.sdk.testing.TestPipeline;
+import org.apache.beam.sdk.transforms.Flatten;
+import org.apache.beam.sdk.transforms.MapElements;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.transforms.SerializableFunction;
+import org.apache.beam.sdk.transforms.SimpleFunction;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.PCollectionList;
+import org.apache.beam.sdk.values.PCollectionTuple;
+import org.apache.beam.sdk.values.TupleTag;
+import org.checkerframework.checker.nullness.qual.NonNull;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.junit.MockitoJUnitRunner;
+
+/** Test class for {@link AccumulatingTableReader}. */
+@RunWith(MockitoJUnitRunner.class)
+public class AccumulatingTableReaderTest implements Serializable {
+  @Rule public final transient TestPipeline testPipeline = TestPipeline.create();
+
+  @Test
+  public void testAccumulatingTableReader() {
+    final String dbName = "testDB";
+    final SourceSchemaReference sourceSchemaReference =
+        SourceSchemaReference.builder().setDbName(dbName).build();
+    final long rowCountPerTable = 10;
+    final long tableCount = 1;
+    final TupleTag<SourceRow> sourceRowTupleTag = new TupleTag<>();
+    final TupleTag<SourceTableReference> sourceTableReferenceTupleTag = new TupleTag<>();
+    ReaderTransformTestUtils readerTransformTestUtils =
+        new ReaderTransformTestUtils(rowCountPerTable, tableCount, sourceSchemaReference);
+    var accumulatingTableReader =
+        readerTransformTestUtils.getTestAccumulatingReader(
+            sourceRowTupleTag, sourceTableReferenceTupleTag);
+    PCollection<TestResultType> results =
+        testPipeline
+            .apply("ReaderUnderTest", accumulatingTableReader)
+            .apply(
+                "result mapper",
+                new TestAccumulatorResultTracker(sourceRowTupleTag, sourceTableReferenceTupleTag));
+
+    PAssert.that(results).satisfies(new TestSatisfyFunction(rowCountPerTable, tableCount));
+    testPipeline.run();
+  }
+
+  enum TestResultType {
+    SOURCE_ROW,
+    TABLE_COMPLETION
+  }
+
+  static class TestAccumulatorResultTracker
+      extends PTransform<@NonNull PCollectionTuple, @NonNull PCollection<TestResultType>> {
+    private final TupleTag<SourceRow> sourceRowTupleTag;
+    private final TupleTag<SourceTableReference> sourceTableReferenceTupleTag;
+
+    TestAccumulatorResultTracker(
+        TupleTag<SourceRow> sourceRowTupleTag,
+        TupleTag<SourceTableReference> sourceTableReferenceTupleTag) {
+      this.sourceRowTupleTag = sourceRowTupleTag;
+      this.sourceTableReferenceTupleTag = sourceTableReferenceTupleTag;
+    }
+
+    @Override
+    @NonNull
+    public PCollection<TestResultType> expand(@NonNull PCollectionTuple input) {
```

Comment:
Can we not do this verification by fetching the tuple elements and verifying counts (using inbuilt count methods) of the 2 pcollections, rather than mapping them to the enum and counting elementwise? Feels more readable that way?

Reply:
Makes a ton of sense.

Reply:
Fixed in followup https://github.com/GoogleCloudPlatform/DataflowTemplates/pull/1449 as this got merged.